zerver/models.py

```diff
@@ -4458,3 +4458,16 @@ def is_authenticated(self) -> bool:
         request.user.is_authenticated verifications.
         """
         return True
+
+
+class StreamTopic(models.Model):
+    # Settings for a Topic. Ideally this eventually won't need to be associated
+    # with a stream, but needs to be as long as there's no Topic table.
+    id: int = models.AutoField(auto_created=True, primary_key=True, verbose_name="ID")
+    realm: Realm = models.ForeignKey(Realm, db_index=True, on_delete=CASCADE)
+    stream: Stream = models.ForeignKey(Stream, on_delete=CASCADE)
+    topic_name: str = models.CharField(max_length=MAX_TOPIC_NAME_LENGTH)
```

Comment:
Oh, I guess one open question this should be `topic`, `topic_name`, or `name`.  Might be good to have a #backend thread discussion about it -- I can see a few tradeoffs. I expect we'll probably want a local variable convention of the form `topic = StreamTopic.objects.get(...)`, and I can imagine `topic.name` feeling like nice syntax, but this needs discussion.

Also I guess we can also take the opportunity to debate whether we want this to be just called `Topic`; I think the current design is that this IS the topic table; it's just the case that rows that don't exist implicitly have `pinned=False`.

Reply:
Recalling a bit more, I think probably the reason we were considering a `StreamTopic` name was to contrast with potential future support for topics in private messages. Given that context ... I think a better model might be to replace `Stream` with `Recipient`, i.e. the constructor would be `Topic(id, realm, recipient, name, pinned=True)`.

Reply:
> Also I guess we can also take the opportunity to debate whether we want this to be just called Topic; I think the current design is that this IS the topic table; it's just the case that rows that don't exist implicitly have pinned=False.

Yeah I think it makes sense to call it the topic table! Thanks. I didn't see this until after talking to Greg, but he had the same thought.

Reply:
And if we call it `Topic`, I think `name` is nicer.

Reply:
> I think a better model might be to replace Stream with Recipient

How difficult would it be to refactor this to `Recipient` once it becomes relevant? It feels like using `Recipient` now would be a more technically complicated, and if we end up using topic IDs before adding support for pinning private messages, or decide to implement pinned private messages in a different way, then this will have been unnecessarily complex. That being said, if doing the migration later will be quite a bit of work then I'm down to set it up that way now.

Reply:
In general it's a pain to rearrange database tables in production (totally doable, and we do it regularly, but like 100x as much work as changing it during development). It's a bit messier to use `Recipient`, but we already use that table in all code paths involving sending and fetching messages, and there's `stream.recipient`, so I don't think it should involve much complexity.

Just beware of fetching a Django object via `Stream.objects.get()` and then accessing its `.reciipent.foo` fields; that'll do extra database queries.  Our helpers like `access_stream_...` will take care of the doing the appropriate `.select_related()` to make sure we fetch the associated Recipient object when we fetch a stream, and one wants to use those for security reasons anyway.